File: src/surface-generation/biomes/random.ts
import * as Mth from "./mth"
import {
    unsignedShift64,
    clamp64,
    rotateLeft64,
    toUnsignedLong,
    toLong,
    toInt,
    remainderUnsigned32,
    fromBytes64,
} from "./mth"
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
import md5 from "md5"

export function toResourceLocation(path: string): string {
    const namespace = "minecraft"

    return namespace + ":" + path
}

export abstract class PositionalRandomFactory {
    abstract fromHashOf(s: string): RandomSource
    abstract at(x: number, y: number, z: number): RandomSource
}

// abstract

export abstract class RandomSource {
    abstract fork(): RandomSource
    abstract forkPositional(): PositionalRandomFactory

    abstract setSeed(seed: bigint): void

    abstract nextInt(bound?: number): number
    abstract nextLong(): bigint
    abstract nextBoolean(): boolean
    abstract nextFloat(): number
    abstract nextDouble(): number
    abstract nextGaussian(): number
}

export class Seed128bit {
    constructor(readonly seedLo: bigint, readonly seedHi: bigint) {}
}

export class RandomSupport {
    public static readonly GOLDEN_RATIO_64 = -7046029254386353131n
    public static readonly SILVER_RATIO_64 = 7640891576956012809n
    private static SEED_UNIQUIFIER = 8682522807148012n

    public static mixStafford13(seed: bigint): bigint {
        seed = clamp64((seed ^ unsignedShift64(seed, 30n)) * -4658895280553007687n)
        seed = clamp64((seed ^ unsignedShift64(seed, 27n)) * -7723592293110705685n)
        return clamp64(seed ^ unsignedShift64(seed, 31n))
    }

    public static upgradeSeedTo128bit(seed: bigint): Seed128bit {
        const lo = seed ^ RandomSupport.SILVER_RATIO_64
        const hi = clamp64(lo + RandomSupport.GOLDEN_RATIO_64)
        return new Seed128bit(RandomSupport.mixStafford13(lo), RandomSupport.mixStafford13(hi))
    }

    public static seedUniquifier(): bigint {
        RandomSupport.SEED_UNIQUIFIER = clamp64(
            RandomSupport.SEED_UNIQUIFIER * 1181783497276652981n
        )

        return RandomSupport.SEED_UNIQUIFIER ^ (BigInt(performance.now()) * 1000000n)
    }

    public static nextIntBetweenInclusive(
        randomSource: RandomSource,
        min: number,
        max: number
    ): number {
        return randomSource.nextInt(max - min + 1) + min
    }

    public static consumeCount(randomSource: RandomSource, count: number): void {
        for (let i = 0; i < count; ++i) {
            randomSource.nextInt()
        }
    }
}

// implementations

export class Xoroshiro128PlusPlus {
    private seedLo: bigint
    private seedHi: bigint

    constructor(lo: bigint | Seed128bit, hi?: bigint) {
        if (lo instanceof Seed128bit) {
            hi = lo.seedHi
            lo = lo.seedLo
        }

        if (hi === undefined) {
            throw new Error()
        }

        this.seedLo = lo
        this.seedHi = hi

        if ((this.seedLo | this.seedHi) == 0n) {
            this.seedLo = RandomSupport.GOLDEN_RATIO_64
            this.seedHi = RandomSupport.SILVER_RATIO_64
        }
    }

    public nextLong(): bigint {
        const i = this.seedLo
        let j = this.seedHi
        const k = clamp64(rotateLeft64(clamp64(i + j), 17n) + i)
        j ^= i
        this.seedLo = rotateLeft64(i, 49n) ^ j ^ clamp64(j << 21n)
        this.seedHi = rotateLeft64(j, 28n)
        return k
    }
}

export class MarsagliaPolarGaussian {
    private nextNextGaussian: number
    private haveNextNextGaussian: boolean

    constructor(readonly randomSource: RandomSource) {}

    public reset(): void {
        this.haveNextNextGaussian = false
    }

    public nextGaussian(): number {
        if (this.haveNextNextGaussian) {
            this.haveNextNextGaussian = false
            return this.nextNextGaussian
        } else {
            // eslint-disable-next-line no-constant-condition
            while (true) {
                const d0 = 2.0 * this.randomSource.nextDouble() - 1.0
                const d1 = 2.0 * this.randomSource.nextDouble() - 1.0
                const d2 = Mth.square(d0) + Mth.square(d1)
                if (!(d2 >= 1.0)) {
                    if (d2 != 0.0) {
                        const d3 = Math.sqrt((-2.0 * Math.log(d2)) / d2)
                        this.nextNextGaussian = d1 * d3
                        this.haveNextNextGaussian = true
                        return d0 * d3
                    }
                }
            }
        }
    }
}

// source implementations

export class XoroshiroRandomSource implements RandomSource {
    private static readonly FLOAT_UNIT = 5.9604645e-8
    private static readonly DOUBLE_UNIT = 1.1102230246251565e-16
    private randomNumberGenerator: Xoroshiro128PlusPlus
    private readonly gaussianSource = new MarsagliaPolarGaussian(this)

    constructor(lo: bigint, hi?: bigint) {
        if (hi === undefined) {
            this.randomNumberGenerator = new Xoroshiro128PlusPlus(
                RandomSupport.upgradeSeedTo128bit(lo)
            )
        } else {
            this.randomNumberGenerator = new Xoroshiro128PlusPlus(lo, hi)
        }
    }

    public fork(): RandomSource {
        return new XoroshiroRandomSource(
            this.randomNumberGenerator.nextLong(),
            this.randomNumberGenerator.nextLong()
        )
    }

    public forkPositional(): PositionalRandomFactory {
        return new XoroshiroPositionalRandomFactory(
            this.randomNumberGenerator.nextLong(),
            this.randomNumberGenerator.nextLong()
        )
    }

    public setSeed(seed: bigint): void {
        this.randomNumberGenerator = new Xoroshiro128PlusPlus(
            RandomSupport.upgradeSeedTo128bit(seed)
        )
        this.gaussianSource.reset()
    }

    public nextInt(bound?: number): number {
        if (bound !== undefined) {
            let i = toUnsignedLong(this.nextInt())
            let j = clamp64(i * toLong(bound))
            let k = j & 4294967295n
            if (k < toLong(bound)) {
                for (
                    let l = remainderUnsigned32(~bound + 1, bound);
                    k < toLong(l);
                    k = j & 4294967295n
                ) {
                    i = toUnsignedLong(this.nextInt())
                    j = i * toLong(bound)
                }
            }

            const i1 = j >> 32n
            return toInt(i1)
        } else {
            return Number(BigInt.asIntN(32, this.randomNumberGenerator.nextLong()))
        }
    }

    public nextLong(): bigint {
        return this.randomNumberGenerator.nextLong()
    }

    public nextBoolean(): boolean {
        return (this.randomNumberGenerator.nextLong() & 1n) != 0n
    }

    public nextFloat(): number {
        return Number(this.nextBits(24)) * XoroshiroRandomSource.FLOAT_UNIT
    }

    public nextDouble(): number {
        return Number(this.nextBits(53)) * XoroshiroRandomSource.DOUBLE_UNIT
    }

    public nextGaussian(): number {
        return this.gaussianSource.nextGaussian()
    }

    public consumeCount(count: number): void {
        for (let i = 0; i < count; ++i) {
            this.randomNumberGenerator.nextLong()
        }
    }

    private nextBits(bits: number): bigint {
        return unsignedShift64(this.randomNumberGenerator.nextLong(), toLong(64 - bits))
    }
}

class XoroshiroPositionalRandomFactory extends PositionalRandomFactory {
    constructor(private readonly seedLo: bigint, private readonly seedHi: bigint) {
        super()
    }

    public at(x: number, y: number, z: number): RandomSource {
        const coordinateSeed = Mth.getSeed(x, y, z)
        const finalSeedLo = coordinateSeed ^ this.seedLo
        return new XoroshiroRandomSource(finalSeedLo, this.seedHi)
    }

    public fromHashOf(s: string): RandomSource {
        const abyte = md5(s, {
            encoding: "utf8",
            asBytes: true,
        })
        const i = fromBytes64(
            abyte[0],
            abyte[1],
            abyte[2],
            abyte[3],
            abyte[4],
            abyte[5],
            abyte[6],
            abyte[7]
        )
        const j = fromBytes64(
            abyte[8],
            abyte[9],
            abyte[10],
            abyte[11],
            abyte[12],
            abyte[13],
            abyte[14],
            abyte[15]
        )
        return new XoroshiroRandomSource(i ^ this.seedLo, j ^ this.seedHi)
    }
}

export abstract class BitRandomSource implements RandomSource {
    private static readonly FLOAT_MULTIPLIER = 5.9604645e-8
    private static readonly DOUBLE_MULTIPLIER = 1.1102230246251565e-16

    abstract next(bits: number): number

    abstract fork(): RandomSource
    abstract forkPositional(): PositionalRandomFactory

    abstract setSeed(seed: bigint): void

    abstract nextGaussian(): number

    nextInt(bound?: number): number {
        if (bound !== undefined) {
            if ((bound & (bound - 1)) == 0) {
                return toInt((toLong(bound) * toLong(this.next(31))) >> 31n)
            } else {
                let i, clapmedI
                do {
                    i = this.next(31)
                    clapmedI = i % bound
                } while (i - clapmedI + (bound - 1) < 0)

                return clapmedI
            }
        } else {
            return this.next(32)
        }
    }

    nextLong(): bigint {
        const i = this.next(32)
        const j = this.next(32)
        const k = toLong(i) << 32n
        return clamp64(k + toLong(j))
    }

    nextBoolean(): boolean {
        return this.next(1) != 0
    }

    nextFloat(): number {
        return this.next(24) * BitRandomSource.FLOAT_MULTIPLIER
    }

    nextDouble(): number {
        const i = this.next(26)
        const j = this.next(27)
        const k = clamp64((toLong(i) << 27n) + toLong(j))
        return Number(k) * BitRandomSource.DOUBLE_MULTIPLIER
    }
}

export class LegacyRandomSource extends BitRandomSource {
    private static readonly MODULUS_BITS = 48n
    private static readonly MODULUS_MASK = 281474976710655n
    private static readonly MULTIPLIER = 25214903917n
    private static readonly INCREMENT = 11n
    private seed: bigint
    private readonly gaussianSource = new MarsagliaPolarGaussian(this)

    constructor(seed: bigint) {
        super()
        this.setSeed(seed)
    }

    public fork(): RandomSource {
        return new LegacyRandomSource(this.nextLong())
    }

    public forkPositional(): PositionalRandomFactory {
        return new LegacyPositionalRandomFactory(this.nextLong())
    }

    public setSeed(seed: bigint): void {
        this.seed = (seed ^ LegacyRandomSource.MULTIPLIER) & LegacyRandomSource.MODULUS_MASK
        this.gaussianSource.reset()
    }

    public next(bits: number): number {
        this.seed = clamp64(
            (this.seed * LegacyRandomSource.MULTIPLIER + LegacyRandomSource.INCREMENT) &
                LegacyRandomSource.MODULUS_MASK
        )
        return toInt(this.seed >> (LegacyRandomSource.MODULUS_BITS - toLong(bits)))
    }

    public nextGaussian(): number {
        return this.gaussianSource.nextGaussian()
    }
}

export function hashCode(s: string): number {
    let h = 0
    for (let i = 0; i < s.length; i++) h = (Math.imul(31, h) + s.charCodeAt(i)) | 0
    return h
}

class LegacyPositionalRandomFactory extends PositionalRandomFactory {
    constructor(private readonly seed: bigint) {
        super()
    }

    public at(x: number, y: number, z: number): RandomSource {
        const coordinateSeed = Mth.getSeed(x, y, z)
        const finalSeed = coordinateSeed ^ this.seed
        return new LegacyRandomSource(finalSeed)
    }

    public fromHashOf(s: string): RandomSource {
        const i = hashCode(s)
        return new LegacyRandomSource(toLong(i) ^ this.seed)
    }
}

// random

class Random {
    private static readonly multiplier = 0x5deece66dn
    private static readonly addend = 0xbn
    private static readonly mask = (1n << 48n) - 1n

    private static readonly FLOAT_UNIT = 1.0 / Number(1n << 24n)
    private static readonly DOUBLE_UNIT = 1.0 / Number(1n << 53n)

    private seed: bigint

    private nextNextGaussian: number
    private haveNextNextGaussian = false

    constructor(seed: bigint) {
        this.setSeed(seed)
    }

    public setSeed(seed: bigint): void {
        this.seed = Random.initialScramble(seed)
        this.haveNextNextGaussian = false
    }

    private static initialScramble(seed: bigint): bigint {
        return (seed ^ Random.multiplier) & Random.mask
    }

    next(bits: number): number {
        const oldseed = this.seed
        const nextseed = (oldseed * Random.multiplier + Random.addend) & Random.mask
        this.seed = nextseed
        return toInt(unsignedShift64(nextseed, toLong(48 - bits)))
    }

    nextInt(bound?: number): number {
        if (bound === undefined) {
            return this.next(32)
        } else {
            let r = this.next(31)
            const m = bound - 1
            if ((bound & m) == 0)
                // i.e., bound is a power of 2
                r = toInt((toLong(bound) * toLong(r)) >> 31n)
            else {
                // reject over-represented candidates
                for (let u = r; u - (r = u % bound) + m < 0; u = this.next(31));
            }
            return r
        }
    }

    nextLong(): bigint {
        // it's okay that the bottom word remains signed.
        return (toLong(this.next(32)) << 32n) + toLong(this.next(32))
    }

    nextBoolean(): boolean {
        return this.next(1) != 0
    }

    nextFloat(): number {
        return this.next(24) / Random.FLOAT_UNIT
    }

    nextDouble(): number {
        return Number((toLong(this.next(26)) << 27n) + toLong(this.next(27))) * Random.DOUBLE_UNIT
    }

    nextGaussian(): number {
        // See Knuth, TAOCP, Vol. 2, 3rd edition, Section 3.4.1 Algorithm C.
        if (this.haveNextNextGaussian) {
            this.haveNextNextGaussian = false
            return this.nextNextGaussian
        } else {
            let v1, v2, s
            do {
                v1 = 2 * this.nextDouble() - 1 // between -1 and 1
                v2 = 2 * this.nextDouble() - 1 // between -1 and 1
                s = v1 * v1 + v2 * v2
            } while (s >= 1 || s == 0)
            const multiplier = Math.sqrt((-2 * Math.log(s)) / s)
            this.nextNextGaussian = v2 * multiplier
            this.haveNextNextGaussian = true
            return v1 * multiplier
        }
    }
}

// world gen
export class WorldgenRandom extends Random {
    private count: number

    constructor(private readonly randomSource: RandomSource) {
        super(0n)
        this.count = 0
    }

    getCount(): number {
        return this.count
    }

    fork(): RandomSource {
        return this.randomSource.fork()
    }

    forkPositional(): PositionalRandomFactory {
        return this.randomSource.forkPositional()
    }

    next(bits: number): number {
        ++this.count
        const randomsource = this.randomSource
        if (randomsource instanceof LegacyRandomSource) {
            const legacyrandomsource = randomsource
            return legacyrandomsource.next(bits)
        } else {
            return toInt(unsignedShift64(this.randomSource.nextLong(), toLong(64 - bits)))
        }
    }

    setSeed(seed: bigint): void {
        if (this.randomSource != null) {
            this.randomSource.setSeed(seed)
        }
    }

    setDecorationSeed(seed: bigint, x: number, y: number): bigint {
        this.setSeed(seed)
        const xSeed = this.nextLong() | 1n
        const ySeed = this.nextLong() | 1n
        const finalSeed = (clamp64(toLong(x) * xSeed) + clamp64(toLong(y) * ySeed)) ^ seed
        this.setSeed(finalSeed)
        return finalSeed
    }

    setFeatureSeed(seed: bigint, x: number, y: number): void {
        const finalSeed = clamp64(seed + toLong(x) + toLong(toInt(10000n * toLong(y))))
        this.setSeed(finalSeed)
    }

    setLargeFeatureSeed(seed: bigint, x: number, y: number): void {
        this.setSeed(seed)
        const xSeed = this.nextLong()
        const ySeed = this.nextLong()
        const finalSeed = clamp64(toLong(x) * xSeed) ^ clamp64(toLong(y) * ySeed) ^ seed
        this.setSeed(finalSeed)
    }

    setLargeFeatureWithSalt(salt: bigint, x: number, y: number, z: number): void {
        const finalSeed = clamp64(
            clamp64(toLong(x) * 341873128712n) +
                clamp64(toLong(y) * 132897987541n) +
                salt +
                toLong(z)
        )
        this.setSeed(finalSeed)
    }

    public static seedSlimeChunk(x: number, y: number, seed: bigint, salt: bigint): Random {
        return new Random(
            (seed +
                toLong(toInt(toLong(x) * toLong(x) * 4987142n)) +
                toLong(toInt(toLong(x) * 5947611n)) +
                toLong(toInt(toLong(y) * toLong(y))) * 4392871n +
                toLong(toInt(toLong(y) * 389711n))) ^
                salt
        )
    }
}

export enum Algorithm {
    LEGACY,
    XOROSHIRO,
}

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
const AlgorithmClasses: { [key in Algorithm]: new (seed: bigint) => RandomSource } = {}
AlgorithmClasses[Algorithm.LEGACY] = LegacyRandomSource
AlgorithmClasses[Algorithm.XOROSHIRO] = XoroshiroRandomSource

export function Algorithm_newInstance(algorithm: Algorithm, seed: bigint): RandomSource {
    return new AlgorithmClasses[algorithm](seed)
}

File: src/cpp/surface-generation/biomes/random.cpp
#include "random.hpp"

// Utils

int32_t hashCode(string s) {
    int32_t h = 0;
    for (int32_t i = 0; i < s.length(); i++) {
        h = (31 * h) + s[i];
    }
    return h;
}

string toResourceLocation(string path) {
    string _namespace = "minecraft";
    return _namespace + ":" + path;
}

// RandomSource

int32_t RandomSource::nextIntBetweenInclusive(int32_t min, int32_t max) {
    return this->nextInt(max - min + 1) + min;
}

void RandomSource::consumeCount(int32_t count) {
    for (int32_t i = 0; i < count; ++i) {
        this->nextInt();
    }
}

// PositionalRandomFactory

unique_ptr<RandomSource> PositionalRandomFactory::at(BlockPos const &pos) {
    return this->at(pos.getX(), pos.getY(), pos.getZ());
}

unique_ptr<RandomSource> PositionalRandomFactory::fromHashOfResourceLocation(string loc) {
    return this->fromHashOf(toResourceLocation(loc));
}

// MarsagliaPolarGaussian

MarsagliaPolarGaussian::MarsagliaPolarGaussian(RandomSource &randomSource) : randomSource(randomSource) {
}

void MarsagliaPolarGaussian::reset() {
    this->haveNextNextGaussian = false;
}

double MarsagliaPolarGaussian::nextGaussian() {
    if (this->haveNextNextGaussian) {
        this->haveNextNextGaussian = false;
        return this->nextNextGaussian;
    } else {
        while (true) {
            double d0 = 2.0 * this->randomSource.nextDouble() - 1.0;
            double d1 = 2.0 * this->randomSource.nextDouble() - 1.0;
            double d2 = Mth::square(d0) + Mth::square(d1);
            if (!(d2 >= 1.0)) {
                if (d2 != 0.0) {
                    double d3 = sqrt(-2.0 * log(d2) / d2);
                    this->nextNextGaussian = d1 * d3;
                    this->haveNextNextGaussian = true;
                    return d0 * d3;
                }
            }
        }
    }
}

// Xoroshiro128PlusPlus

Xoroshiro128PlusPlus::Xoroshiro128PlusPlus(RandomSupport::Seed128bit const &seed)
    : Xoroshiro128PlusPlus(seed.seedLo, seed.seedHi) {
}

Xoroshiro128PlusPlus::Xoroshiro128PlusPlus(int64_t seedLo, int64_t seedHi) {
    this->seedLo = seedLo;
    this->seedHi = seedHi;
    if ((this->seedLo | this->seedHi) == 0LL) {
        this->seedLo = -7046029254386353131LL;
        this->seedHi = 7640891576956012809LL;
    }
}

int64_t Xoroshiro128PlusPlus::nextLong() {
    int64_t seedLo = this->seedLo;
    int64_t seedHi = this->seedHi;
    int64_t rotatedSeed = rotateLeft(seedLo + seedHi, 17) + seedLo;
    seedHi ^= seedLo;
    this->seedLo = rotateLeft(seedLo, 49) ^ seedHi ^ (seedHi << 21);
    this->seedHi = rotateLeft(seedHi, 28);
    return rotatedSeed;
}

// XoroshiroRandomSource

XoroshiroRandomSource::XoroshiroRandomSource(int64_t seed)
    : randomNumberGenerator(RandomSupport::upgradeSeedTo128bit(seed)) {
    objectCreated("RandomSource");
}

XoroshiroRandomSource::XoroshiroRandomSource(int64_t seedLo, int64_t seedHi)
    : randomNumberGenerator(Xoroshiro128PlusPlus(seedLo, seedHi)) {
    objectCreated("RandomSource");
}

unique_ptr<RandomSource> XoroshiroRandomSource::fork() {
    int64_t seedLo = this->randomNumberGenerator.nextLong();
    int64_t seedHi = this->randomNumberGenerator.nextLong();
    return make_unique<XoroshiroRandomSource>(seedLo, seedHi);
}

unique_ptr<PositionalRandomFactory> XoroshiroRandomSource::forkPositional() {
    int64_t seedLo = this->randomNumberGenerator.nextLong();
    int64_t seedHi = this->randomNumberGenerator.nextLong();
    return make_unique<XoroshiroRandomSource::XoroshiroPositionalRandomFactory>(seedLo, seedHi);
}

void XoroshiroRandomSource::setSeed(int64_t seed) {
    this->randomNumberGenerator = Xoroshiro128PlusPlus(RandomSupport::upgradeSeedTo128bit(seed));
    this->gaussianSource.reset();
}

int32_t XoroshiroRandomSource::nextInt() {
    return (int32_t)this->randomNumberGenerator.nextLong();
}

int32_t XoroshiroRandomSource::nextInt(int32_t bound) {
    int64_t i = toUnsignedLong(this->nextInt());
    int64_t j = i * (int64_t)bound;
    int64_t k = j & 4294967295LL;
    if (k < (int64_t)bound) {
        for (int32_t l = remainderUnsigned(~bound + 1, bound); k < (int64_t)l; k = j & 4294967295LL) {
            i = toUnsignedLong(this->nextInt());
            j = i * (int64_t)bound;
        }
    }

    int64_t i1 = j >> 32;
    return (int32_t)i1;
}

int64_t XoroshiroRandomSource::nextLong() {
    return this->randomNumberGenerator.nextLong();
}

bool XoroshiroRandomSource::nextBoolean() {
    return (this->randomNumberGenerator.nextLong() & 1LL) != 0LL;
}

float XoroshiroRandomSource::nextFloat() {
    return (float)this->nextBits(24) * 5.9604645E-8F;
}

double XoroshiroRandomSource::nextDouble() {
    return (double)this->nextBits(53) * (double)1.110223E-16F;
}

double XoroshiroRandomSource::nextGaussian() {
    return this->gaussianSource.nextGaussian();
}

void XoroshiroRandomSource::consumeCount(int32_t count) {
    for (int32_t i = 0; i < count; ++i) {
        this->randomNumberGenerator.nextLong();
    }
}

int64_t XoroshiroRandomSource::nextBits(int32_t bits) {
    return ushr_l(this->randomNumberGenerator.nextLong(), (64 - bits));
}

// XoroshiroPositionalRandomFactory

XoroshiroRandomSource::XoroshiroPositionalRandomFactory::XoroshiroPositionalRandomFactory(int64_t seedLo,
                                                                                          int64_t seedHi)
    : seedLo(seedLo), seedHi(seedHi) {
    objectCreated("PositionalRandomFactory");
}

unique_ptr<RandomSource> XoroshiroRandomSource::XoroshiroPositionalRandomFactory::at(int32_t x, int32_t y, int32_t z) {
    int64_t i = Mth::getSeed(x, y, z);
    int64_t j = i ^ this->seedLo;
    return make_unique<XoroshiroRandomSource>(j, this->seedHi);
}

unique_ptr<RandomSource> XoroshiroRandomSource::XoroshiroPositionalRandomFactory::fromHashOf(string s) {
    MD5_CTX md5;
    BYTE hash[MD5_BLOCK_SIZE];

    md5_init(&md5);
    md5_update(&md5, (const BYTE *)s.c_str(), s.length() * sizeof(char));
    md5_final(&md5, hash);

    int64_t lo = fromBytes(hash[0], hash[1], hash[2], hash[3], hash[4], hash[5], hash[6], hash[7]);
    int64_t hi = fromBytes(hash[8], hash[9], hash[10], hash[11], hash[12], hash[13], hash[14], hash[15]);
    return make_unique<XoroshiroRandomSource>(lo ^ this->seedLo, hi ^ this->seedHi);
}

// BitRandomSource

int32_t BitRandomSource::nextInt() {
    return this->next(32);
}

int32_t BitRandomSource::nextInt(int32_t bound) {
    if ((bound & bound - 1) == 0) {
        return (int32_t)((int64_t)bound * (int64_t)this->next(31) >> 31);
    } else {
        int32_t i;
        int32_t j;
        do {
            i = this->next(31);
            j = i % bound;
        } while (i - j + (bound - 1) < 0);

        return j;
    }
}

int64_t BitRandomSource::nextLong() {
    int32_t i = this->next(32);
    int32_t j = this->next(32);
    int64_t k = (int64_t)i << 32;
    return k + (int64_t)j;
}

bool BitRandomSource::nextBoolean() {
    return this->next(1) != 0;
}

float BitRandomSource::nextFloat() {
    return (float)this->next(24) * FLOAT_MULTIPLIER;
}

double BitRandomSource::nextDouble() {
    int32_t i = this->next(26);
    int32_t j = this->next(27);
    int64_t k = ((int64_t)i << 27) + (int64_t)j;
    return (double)k * DOUBLE_MULTIPLIER;
}

// LegacyRandomSource

LegacyRandomSource::LegacyRandomSource(int64_t seed) {
    this->setSeed(seed);
    objectCreated("RandomSource");
}

unique_ptr<RandomSource> LegacyRandomSource::fork() {
    return make_unique<LegacyRandomSource>(this->nextLong());
}

unique_ptr<PositionalRandomFactory> LegacyRandomSource::forkPositional() {
    return make_unique<LegacyRandomSource::LegacyPositionalRandomFactory>(this->nextLong());
}

void LegacyRandomSource::setSeed(int64_t seed) {
    this->seed = (seed ^ MULTIPLIER) & MODULUS_MASK;
    this->gaussianSource.reset();
}

int32_t LegacyRandomSource::next(int32_t bits) {
    int64_t seed = this->seed;
    int64_t newSeed = seed * MULTIPLIER + INCREMENT & MODULUS_MASK;
    this->seed = newSeed;

    return (int32_t)(newSeed >> (MODULUS_BITS - bits));
}

double LegacyRandomSource::nextGaussian() {
    return this->gaussianSource.nextGaussian();
}

// LegacyPositionalRandomFactory

LegacyRandomSource::LegacyPositionalRandomFactory::LegacyPositionalRandomFactory(int64_t seed) : seed(seed) {
    objectCreated("PositionalRandomFactory");
}

unique_ptr<RandomSource> LegacyRandomSource::LegacyPositionalRandomFactory::at(int32_t x, int32_t y, int32_t z) {
    int64_t seed = Mth::getSeed(x, y, z);
    int64_t newSeed = seed ^ this->seed;
    return make_unique<LegacyRandomSource>(newSeed);
}

unique_ptr<RandomSource> LegacyRandomSource::LegacyPositionalRandomFactory::fromHashOf(string s) {
    int32_t hash = hashCode(s);
    return make_unique<LegacyRandomSource>((int64_t)hash ^ this->seed);
}

// Random

Random::Random(int64_t seed) {
    this->setSeed(seed);
}

unique_ptr<RandomSource> Random::fork() {
    return nullptr;
}

unique_ptr<PositionalRandomFactory> Random::forkPositional() {
    return nullptr;
}

void Random::setSeed(int64_t seed) {
    this->seed = initialScramble(seed);
    this->haveNextNextGaussian = false;
}

int32_t Random::next(int32_t bits) {
    this->seed = (this->seed * multiplier + addend) & mask;
    return (int32_t)(ushr_l(this->seed, (48 - bits)));
}

int32_t Random::nextInt() {
    return next(32);
}

int32_t Random::nextInt(int32_t bound) {
    int32_t r = next(31);
    int32_t m = bound - 1;
    if ((bound & m) == 0) // i.e., bound is a power of 2
        r = (int32_t)((bound * (int64_t)r) >> 31);
    else { // reject over-represented candidates
        for (int32_t u = r; u - (r = u % bound) + m < 0; u = next(31))
            ;
    }
    return r;
}

int64_t Random::nextLong() {
    // it's okay that the bottom word remains signed.
    return ((int64_t)(next(32)) << 32) + next(32);
}

bool Random::nextBoolean() {
    return next(1) != 0;
}

float Random::nextFloat() {
    return next(24) / ((float)(1 << 24));
}

double Random::nextDouble() {
    return (((int64_t)(next(26)) << 27) + next(27)) * DOUBLE_UNIT;
}

double Random::nextGaussian() {
    // See Knuth, TAOCP, Vol. 2, 3rd edition, Section 3.4.1 Algorithm C.
    if (haveNextNextGaussian) {
        haveNextNextGaussian = false;
        return nextNextGaussian;
    } else {
        double v1, v2, s;
        do {
            v1 = 2 * nextDouble() - 1; // between -1 and 1
            v2 = 2 * nextDouble() - 1; // between -1 and 1
            s = v1 * v1 + v2 * v2;
        } while (s >= 1 || s == 0);
        double multiplier = sqrt(-2 * log(s) / s);
        nextNextGaussian = v2 * multiplier;
        haveNextNextGaussian = true;
        return v1 * multiplier;
    }
}

// WorldgenRandom

WorldgenRandom::WorldgenRandom(unique_ptr<RandomSource> randomSource)
    : Random(0LL), randomSource(std::move(randomSource)), count(0) {
    objectCreated("RandomSource");
}

unique_ptr<RandomSource> WorldgenRandom::fork() {
    return this->randomSource->fork();
}

unique_ptr<PositionalRandomFactory> WorldgenRandom::forkPositional() {
    return this->randomSource->forkPositional();
}

void WorldgenRandom::setSeed(int64_t seed) {
    if (this->randomSource != nullptr) {
        this->randomSource->setSeed(seed);
    }
}

int32_t WorldgenRandom::getCount() {
    return this->count;
}

int32_t WorldgenRandom::next(int32_t bits) {
    ++this->count;
    RandomSource *randomsource = this->randomSource.get();
    if (LegacyRandomSource *legacyrandomsource = dynamic_cast<LegacyRandomSource *>(randomsource)) {
        return legacyrandomsource->next(bits);
    } else {
        return (int32_t)(ushr_l(this->randomSource->nextLong(), (64 - bits)));
    }
}

int64_t WorldgenRandom::setDecorationSeed(int64_t seed, int32_t x, int32_t y) {
    this->setSeed(seed);
    int64_t i = this->nextLong() | 1LL;
    int64_t j = this->nextLong() | 1LL;
    int64_t newSeed = (int64_t)x * i + (int64_t)y * j ^ seed;
    this->setSeed(newSeed);
    return newSeed;
}

void WorldgenRandom::setFeatureSeed(int64_t seed, int32_t counter1, int32_t counter2) {
    int64_t newSeed = seed + (int64_t)counter1 + (int64_t)(10000 * counter2);
    this->setSeed(newSeed);
}

void WorldgenRandom::setLargeFeatureSeed(int64_t seed, int32_t x, int32_t y) {
    this->setSeed(seed);
    int64_t i = this->nextLong();
    int64_t j = this->nextLong();
    int64_t newSeed = (int64_t)x * i ^ (int64_t)y * j ^ seed;
    this->setSeed(newSeed);
}

void WorldgenRandom::setLargeFeatureWithSalt(int64_t salt, int32_t x, int32_t y, int32_t z) {
    int64_t seed = (int64_t)x * 341873128712LL + (int64_t)y * 132897987541LL + salt + (int64_t)z;
    this->setSeed(seed);
}

unique_ptr<Random> WorldgenRandom::seedSlimeChunk(int32_t x, int32_t z, int64_t seed, int64_t salt) {
    return make_unique<Random>(seed + (int64_t)(x * x * 4987142) + (int64_t)(x * 5947611) +
                                   (int64_t)(z * z) * 4392871LL + (int64_t)(z * 389711) ^
                               salt);
}

unique_ptr<RandomSource> WorldgenRandom::Algorithm_newInstance(Algorithm algorithm, int64_t seed) {
    if (algorithm == Algorithm::LEGACY) {
        return make_unique<LegacyRandomSource>(seed);
    } else if (algorithm == Algorithm::XOROSHIRO) {
        return make_unique<XoroshiroRandomSource>(seed);
    } else {
        return nullptr;
    }
}